core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java

```diff
@@ -0,0 +1,183 @@
+package org.infinispan.executors;
+
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Executes tasks in the given executor, but never has more than {@code maxConcurrentTasks} tasks running at the same time.
+ *
+ * @author Dan Berindei
+ * @since 7.2
+ */
+public class SemaphoreCompletionService<T> implements CompletionService<T> {
+   private static final Log log = LogFactory.getLog(SemaphoreCompletionService.class);
+   // Trace logs are always disabled by default
+   private static final boolean trace = false;
```

Comment:
I was thinking none of the existing executors log anything, so I'd make this the same. But I see `BlockingTaskAwareExecutorImpl` does have some trace logs, so I'll change it.